identity-actor/src/actor/actor.rs

```diff
@@ -0,0 +1,279 @@
+      }
+      None => Err(RemoteSendError::UnexpectedRequest(endpoint.to_string())),
+    }
+  }
+
+  pub fn commander(&mut self) -> &mut NetCommander {
+    &mut self.commander
+  }
+
+  /// Start listening on the given `address`. Returns the first address that the actor started listening on, which may
+  /// be different from `address` itself, e.g. when passing addresses like `/ip4/0.0.0.0/tcp/0`. Even when passing a
+  /// single address, multiple addresses may end up being listened on. To obtain all those addresses, use
+  /// [`Actor::addresses`]. Note that even when the same address is passed, the returned address is not deterministic,
+  /// and should thus not be relied upon.
+  pub async fn start_listening(&mut self, address: Multiaddr) -> ActorResult<Multiaddr> {
+    self.commander().start_listening(address).await
+  }
+
+  /// Return all addresses that are currently being listened on.
+  pub async fn addresses(&mut self) -> ActorResult<Vec<Multiaddr>> {
+    self.commander().get_addresses().await
+  }
+
+  /// Shut this actor down. This will break the event loop in the background immediately,
+  /// returning an error for all current handlers that interact with their copy of the
+  /// actor or those waiting on messages. The actor will thus stop listening on all addresses.
+  ///
+  /// Calling this and other methods, which interact with the event loop, on an actor that was shutdown
+  /// will return [`Error::Shutdown`].
+  pub async fn shutdown(mut self) -> ActorResult<()> {
+    // Consuming self drops the internal commander. If this is the last copy of the commander,
+    // the event loop will break as a result. However, if copies exist, such as in running handlers,
+    // this function will return while the event loop keeps running. Ideally we could then join on the background task
+    // to wait for all handlers to finish gracefully. However, not all spawn functions return a JoinHandle,
+    // such as wasm_bindgen_futures::spawn_local. The current alternative is to use a non-graceful exit,
+    // which breaks the event loop immediately and returns an error through all open channels that require a result.
+    self.commander().shutdown().await
+  }
+
+  /// Associate the given `peer_id` with an `address`. This needs to be done before sending a
+  /// request to this [`PeerId`].
+  pub async fn add_address(&mut self, peer_id: PeerId, address: Multiaddr) -> ActorResult<()> {
+    self.commander().add_addresses(peer_id, OneOrMany::One(address)).await
+  }
+
+  /// Associate the given `peer_id` with multiple `addresses`. This needs to be done before sending a
+  /// request to this [`PeerId`].
+  pub async fn add_addresses(&mut self, peer_id: PeerId, addresses: Vec<Multiaddr>) -> ActorResult<()> {
+    self
+      .commander()
+      .add_addresses(peer_id, OneOrMany::Many(addresses))
+      .await
+  }
+
+  /// Sends a synchronous request to a peer and returns its response.
+  pub async fn send_request<REQ: SyncActorRequest>(
```

Comment:
It might be useful to explain the reason why this method is or needs to be async given that the request itself is synchronous.

Reply:
Thanks for this comment. It shows that the current language around sync and async actor requests is not ideal, because devs tend to think about it in terms of sync/async functions in Rust.
The synchronicity of an actor request is a higher-level concept than that of Rust functions. Such an actor request is still sent using async I/O, and hence the function needs to be async. But conceptually, the request is synchronous in that it wait for the peer's handler to finish execution. Does that make sense?

To my knowledge, sync/async is the most appropriate terminology to use, but it's apparently still confusing. If you have alternative ideas for what language to use here, I'd be happy to change it if it fits.